File: backend-fastapi/services/embedding.py
# ── Embedding Service ────────────────────────────────
# Generates 384-dim vector embeddings via Hugging Face Inference API.
# Model: sentence-transformers/all-MiniLM-L6-v2 (free tier eligible).

import os
import httpx
from typing import List

HF_MODEL_URL = (
    "https://router.huggingface.co/hf-inference/models/"
    "sentence-transformers/all-MiniLM-L6-v2/pipeline/feature-extraction"
)


async def get_embedding(text: str) -> List[float]:
    """
    Generate a 384-dimensional vector embedding for the given text.

    Args:
        text: The text to embed.

    Returns:
        384-dim float array.
    """
    api_key = os.getenv("HF_API_KEY")
    if not api_key:
        raise ValueError("HF_API_KEY is not set in .env — add your Hugging Face API key.")

    # Truncate to ~8000 chars to stay within model token limits
    truncated = text[:8000] if len(text) > 8000 else text

    async with httpx.AsyncClient() as client:
        response = await client.post(
            HF_MODEL_URL,
            headers={
                "Authorization": f"Bearer {api_key}",
                "Content-Type": "application/json",
            },
            json={
                "inputs": truncated,
                "options": {"wait_for_model": True},
            },
            timeout=30.0,
        )

        if not response.is_success:
            raise RuntimeError(
                f"Hugging Face API error ({response.status_code}): {response.text}"
            )

        result = response.json()

    # Handle different response shapes from the HF API:
    # 1. Nested array (token-level embeddings): [[v1], [v2], ...] → mean-pool
    # 2. Single vector wrapped: [[0.1, 0.2, ...]]
    # 3. Flat array: [0.1, 0.2, ...]

    if isinstance(result, list) and len(result) > 0:
        if isinstance(result[0], list):
            if len(result[0]) > 0 and isinstance(result[0][0], (int, float)):
                # Shape: [[0.1, 0.2, ...]] — single vector wrapped in outer array
                return result[0]
            # Token-level embeddings: mean pool across tokens
            num_tokens = len(result)
            dim = len(result[0])
            pooled = [0.0] * dim
            for token_embedding in result:
                for j in range(dim):
                    pooled[j] += token_embedding[j]
            for j in range(dim):
                pooled[j] /= num_tokens
            return pooled

        if isinstance(result[0], (int, float)):
            return result

    raise RuntimeError("Unexpected embedding response format from Hugging Face.")

File: backend/services/embedding.js
// ── Embedding Service ────────────────────────────────
// Generates 384-dim vector embeddings via Hugging Face Inference API.
// Model: sentence-transformers/all-MiniLM-L6-v2 (free tier eligible).

const HF_MODEL_URL =
  "https://router.huggingface.co/hf-inference/models/sentence-transformers/all-MiniLM-L6-v2/pipeline/feature-extraction";

/**
 * Generate a 384-dimensional vector embedding for the given text.
 *
 * @param {string} text - The text to embed.
 * @returns {Promise<number[]>} 384-dim float array.
 */
async function getEmbedding(text) {
  const apiKey = process.env.HF_API_KEY;
  if (!apiKey) {
    throw new Error("HF_API_KEY is not set in .env — add your Hugging Face API key.");
  }

  // Truncate to ~8000 chars to stay within model token limits
  const truncated = text.length > 8000 ? text.slice(0, 8000) : text;
  console.log('inside embedding')
  const response = await fetch(HF_MODEL_URL, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${apiKey}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      inputs: truncated,
      options: { wait_for_model: true },
    }),
  });

  if (!response.ok) {
    const errBody = await response.text();
    throw new Error(`Hugging Face API error (${response.status}): ${errBody}`);
  }

  const result = await response.json();

  // Handle different response shapes from the HF API:
  // 1. Nested array (token-level embeddings): [[v1], [v2], ...] → mean-pool
  // 2. Single vector wrapped: [[0.1, 0.2, ...]]
  // 3. Flat array: [0.1, 0.2, ...]

  if (Array.isArray(result) && Array.isArray(result[0])) {
    if (typeof result[0][0] === "number") {
      // Shape: [[0.1, 0.2, ...]] — single vector wrapped in outer array
      return result[0];
    }
    // Token-level embeddings: mean pool across tokens
    const numTokens = result.length;
    const dim = result[0].length;
    const pooled = new Array(dim).fill(0);
    for (let i = 0; i < numTokens; i++) {
      for (let j = 0; j < dim; j++) {
        pooled[j] += result[i][j];
      }
    }
    for (let j = 0; j < dim; j++) {
      pooled[j] /= numTokens;
    }
    return pooled;
  }

  if (Array.isArray(result) && typeof result[0] === "number") {
    return result;
  }

  throw new Error("Unexpected embedding response format from Hugging Face.");
}

module.exports = { getEmbedding };
